File: sdk/src/pda.ts
/**
 * Every address the program derives, and the bin-array arithmetic that decides
 * which of them a given bin range touches.
 */
import { PublicKey } from "@solana/web3.js";
import { BINS_PER_ARRAY, MAX_BIN_ARRAY_INDEX, MIN_BIN_ARRAY_INDEX, PROGRAM_ID } from "./constants";
import { seedI64, seedU16 } from "./codec";

/**
 * Canonical mint order. The pool PDA is seeded with both mints, so X must sort
 * below Y or the same pair would be creatable at two addresses.
 */
export function comparePublicKeys(a: PublicKey, b: PublicKey) {
  return Buffer.compare(a.toBuffer(), b.toBuffer());
}

/** `[x, y]` in the order the pool PDA requires, whichever way they came in. */
export function orderMints(a: PublicKey, b: PublicKey): [PublicKey, PublicKey] {
  return comparePublicKeys(a, b) < 0 ? [a, b] : [b, a];
}

const pda = (seeds: (Buffer | Uint8Array)[]) => PublicKey.findProgramAddressSync(seeds, PROGRAM_ID)[0];

export const configPda = (authority: PublicKey, index: number) =>
  pda([Buffer.from("config"), authority.toBuffer(), seedU16(index)]);

export const poolPda = (config: PublicKey, mintX: PublicKey, mintY: PublicKey) =>
  pda([Buffer.from("pool"), config.toBuffer(), mintX.toBuffer(), mintY.toBuffer()]);

export const reservePda = (pool: PublicKey, mint: PublicKey) =>
  pda([Buffer.from("reserve"), pool.toBuffer(), mint.toBuffer()]);

export const binArrayPda = (pool: PublicKey, index: number) =>
  pda([Buffer.from("bin_array"), pool.toBuffer(), seedI64(index)]);

// A position has no PDA. It is a plain keypair account, because its band
// moves — `resize_position` changes both edges — and an address derived from a
// band would be stale the moment it did. Positions are found with
// `getProgramAccounts` filtered on the owner, which is how clients have always
// found them; the keypair is needed only to sign the account into existence.

/** Floors towards -infinity, matching `i32::div_euclid` on chain. */
export const binArrayIndex = (binId: number) => Math.floor(binId / BINS_PER_ARRAY);

export const binArrayLower = (index: number) => index * BINS_PER_ARRAY;
export const binArrayUpper = (index: number) => index * BINS_PER_ARRAY + BINS_PER_ARRAY - 1;

export function arrayIndexesFor(lowerBinId: number, upperBinId: number) {
  const lo = binArrayIndex(lowerBinId);
  const hi = binArrayIndex(upperBinId);
  return Array.from({ length: hi - lo + 1 }, (_, i) => lo + i);
}

/**
 * The bin arrays to hand a swap, in the order it will walk them.
 *
 * A swap starts at the active bin and moves down for X→Y or up for Y→X, so the
 * arrays it can reach are the active one and its neighbours in that direction.
 * Only arrays that exist are named: the program takes them as remaining
 * accounts and an address with no account behind it fails the instruction.
 *
 * `reach` is how many arrays past the active one to carry. Three covers 210
 * bins, which is further than a single swap will normally travel, and each
 * extra array is 6,792 bytes of account the transaction has to reference.
 *
 * Compute is the real ceiling on how far the walk gets: about 12,200 CU per
 * funded bin crossed against the 1.4M transaction limit caps any swap at 113
 * bins, and 23 under the 300k a router budgets. Carrying an array the walk
 * never reaches costs only 394 CU, which is why the Jupiter SDK trims to
 * `REACH = 2` and this default does not — see `jupiter/taper-amm-sdk/src/pda.rs`.
 * Both are measured by `swap_cost_by_reach` in `tests/tests/compute.rs`.
 */
export function swapArrayIndexes(
  activeId: number,
  swapForY: boolean,
  exists: (index: number) => boolean,
  reach = 3
) {
  const home = binArrayIndex(activeId);
  const step = swapForY ? -1 : 1;
  const out: number[] = [];
  for (let i = 0; i < reach; i += 1) {
    const index = home + step * i;
    if (index < MIN_BIN_ARRAY_INDEX || index > MAX_BIN_ARRAY_INDEX) break;
    // A missing array *is* a stop: the program breaks its walk the moment it
    // needs an index that was not supplied, so nothing past a gap can trade in
    // this swap. Naming the arrays beyond it is harmless but pointless.
    if (exists(index)) out.push(index);
  }
  return out;
}

File: jupiter/taper-amm-sdk/src/pda.rs
//! Every address the program derives, and which bin arrays a swap needs.
//!
//! Mirrors `sdk/src/pda.ts` seed for seed. The seeds are ABI: an address
//! derived from the wrong ones is not a wrong answer, it is an account that
//! does not exist.

use solana_pubkey::Pubkey;

use crate::TAPER_AMM_ID;
use taper_core::constants::{MAX_BIN_ARRAY_INDEX, MAX_BIN_PER_ARRAY, MIN_BIN_ARRAY_INDEX};

pub use taper_core::math::ladder::{
    bin_array_index, bin_array_lower_bin_id, bin_array_upper_bin_id,
};

fn pda(seeds: &[&[u8]]) -> Pubkey {
    Pubkey::find_program_address(seeds, &TAPER_AMM_ID).0
}

pub fn config_pda(authority: &Pubkey, index: u16) -> Pubkey {
    pda(&[b"config", authority.as_ref(), &index.to_le_bytes()])
}

/// The pool PDA is seeded with both mints in order, so X must sort below Y or
/// the same pair would be creatable at two addresses.
pub fn pool_pda(config: &Pubkey, mint_x: &Pubkey, mint_y: &Pubkey) -> Pubkey {
    pda(&[b"pool", config.as_ref(), mint_x.as_ref(), mint_y.as_ref()])
}

pub fn reserve_pda(pool: &Pubkey, mint: &Pubkey) -> Pubkey {
    pda(&[b"reserve", pool.as_ref(), mint.as_ref()])
}

/// Note the `i64` seed: the index is an `i32` everywhere else, but the seed
/// the program derives with is eight bytes.
pub fn bin_array_pda(pool: &Pubkey, index: i32) -> Pubkey {
    pda(&[b"bin_array", pool.as_ref(), &(index as i64).to_le_bytes()])
}

// A position has no PDA: it is a plain keypair account, because its band moves
// and an address derived from a band would be stale the moment it did. Nothing
// here needs one anyway — a quote reads pools and bin arrays, never positions.

/// How many bin arrays to carry either side of the active one.
///
/// Two: the active bin's array, and one in the direction of travel. Measured
/// in `tests/tests/compute.rs`, a hop costs about 12,200 CU per funded bin it
/// crosses against a 1.4M transaction ceiling, which caps *any* Taper hop at
/// 113 bins — and two arrays cover between 71 and 140 of them, depending on
/// where the active bin sits inside its own array.
///
/// So the third array is not unreachable in principle: with the active bin at
/// the bottom of its own array, two cover only 71 and the budget would stretch
/// to 113. It is unreachable in practice, which is the weaker claim this
/// constant actually needs. A hop crossing 71 funded bins is ~7% of price
/// impact on a 10 bps ladder, which no router quotes, and under the 300k a
/// router really budgets the walk stops at 23 bins — inside the first array in
/// the great majority of cases. Raising this to 3 would buy reach only for
/// hops nothing would route, at 394 CU and 6,792 bytes of account per swap.
pub const REACH: usize = 2;

/// Bins per array, so a caller need not reach into `taper-core` for it.
pub const BINS_PER_ARRAY: usize = MAX_BIN_PER_ARRAY;

fn in_bitmap(index: i32) -> bool {
    (MIN_BIN_ARRAY_INDEX..=MAX_BIN_ARRAY_INDEX).contains(&index)
}

/// The bin arrays a swap may walk, in the order it will walk them.
///
/// `exists` is the pool's own bitmap — `Pool::is_bin_array_occupied` — which
/// is what makes this answerable without a network call. Only arrays that
/// exist are named: an address with no account behind it fails the
/// instruction outright.
///
/// A missing array ends the list rather than being skipped. The program breaks
/// its walk the moment it needs an index it was not handed, so nothing past a
/// gap can trade in this swap and naming it would only cost transaction space.
pub fn swap_array_indexes(
    active_id: i32,
    swap_for_y: bool,
    exists: impl Fn(i32) -> bool,
    reach: usize,
) -> Vec<i32> {
    let home = bin_array_index(active_id);
    let step = if swap_for_y { -1 } else { 1 };
    let mut out = Vec::with_capacity(reach);
    for i in 0..reach as i32 {
        let index = home + step * i;
        if !in_bitmap(index) || !exists(index) {
            break;
        }
        out.push(index);
    }
    out
}

/// Every array a quote might need, in *either* direction.
///
/// `Amm::get_accounts_to_update` runs before the swap direction is known, so
/// it has to fetch both ways: `home - (reach - 1) ..= home + (reach - 1)`.
pub fn arrays_to_fetch(active_id: i32, exists: impl Fn(i32) -> bool, reach: usize) -> Vec<i32> {
    let home = bin_array_index(active_id);
    let span = reach as i32 - 1;
    (home - span..=home + span)
        .filter(|index| in_bitmap(*index) && exists(*index))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_walk_stops_at_the_first_gap() {
        // Arrays 0 and -2 exist, -1 does not. A downward walk from bin 0 can
        // only ever reach array 0.
        let exists = |i: i32| i == 0 || i == -2;
        assert_eq!(swap_array_indexes(0, true, exists, 3), vec![0]);
        // Upward from the same bin there is nothing past the home array.
        assert_eq!(swap_array_indexes(0, false, exists, 3), vec![0]);
    }

    #[test]
    fn a_walk_carries_the_arrays_in_travel_order() {
        let exists = |_: i32| true;
        assert_eq!(swap_array_indexes(0, true, exists, REACH), vec![0, -1]);
        assert_eq!(swap_array_indexes(0, false, exists, REACH), vec![0, 1]);
    }

    #[test]
    fn fetching_covers_both_directions() {
        assert_eq!(arrays_to_fetch(0, |_| true, REACH), vec![-1, 0, 1]);
        // A gap is skipped here rather than ending the list: which side the
        // swap will travel is not known yet, so both are collected.
        assert_eq!(arrays_to_fetch(0, |i| i != -1, REACH), vec![0, 1]);
    }

    #[test]
    fn the_bitmap_bounds_the_reach() {
        let last = MAX_BIN_ARRAY_INDEX;
        let top_bin = bin_array_lower_bin_id(last);
        assert_eq!(
            swap_array_indexes(top_bin, false, |_| true, 3),
            vec![last],
            "nothing exists above the bitmap"
        );
    }
}
